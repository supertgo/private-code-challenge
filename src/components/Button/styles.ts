import styled, { css, DefaultTheme } from 'styled-components';
import { ButtonProps } from '.';

type WrapperProps = Pick<ButtonProps, 'size' | 'color' | 'icon'>;

const wrapperModiffiers = {
  small: () => css`
    width: 13rem;
  `,
  medium: () => css`
    width: 30rem;
  `,
  large: () => css`
    width: 38rem;
  `,

  primary: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
  `,
  secundary: (theme: DefaultTheme) => css`
    color: ${theme.colors.black};
  `,

  gradient: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
  `,
  withIcon: () => css`
    display: flex;
    flex-direction: row;
    justify-content: center;

    span {
      padding-right: 3.1rem;
    }
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, color, size, icon }) => css`
    border: none;
    height: 4.3rem;
    align-items: center;
    text-align: center;
    font-size: ${theme.font.sizes.large};
    font-weight: ${theme.font.bold};
    border-radius: ${theme.border.small};
    background: ${theme.colors[color!]};

    ${icon && size !== 'small' && wrapperModiffiers.withIcon()};
    ${!!size && wrapperModiffiers[size]()};
    ${!!color && wrapperModiffiers[color](theme)};
  `}
`;

export const Icon = styled.div``;
