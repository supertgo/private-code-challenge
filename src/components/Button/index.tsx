import {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes
} from 'hoist-non-react-statics/node_modules/@types/react';
import * as S from './styles';

type ButtonType =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  color?: 'primary' | 'secundary' | 'gradient';
  size?: 'small' | 'medium' | 'large';
  icon?: boolean;
  children: React.ReactNode;
} & ButtonType;

const Button = ({
  color = 'primary',
  size = 'medium',
  icon = false,
  children
}: ButtonProps) => (
  <S.Wrapper color={color} size={size} icon={icon}>
    <span>{children}</span>
    {icon && size != 'small' && (
      <svg
        width="18"
        height="17"
        viewBox="0 0 20 17"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.02165 9.89163H15.8188L11.5959 14.9638C11.4981 15.0814 11.4245 15.2171 11.3791 15.3632C11.3338 15.5093 11.3177 15.6629 11.3318 15.8152C11.3601 16.1229 11.5095 16.4066 11.7471 16.6041C11.9847 16.8015 12.291 16.8965 12.5986 16.8682C12.9062 16.8398 13.19 16.6904 13.3874 16.4528L19.2041 9.47283C19.2432 9.41731 19.2782 9.35899 19.3088 9.29833C19.3088 9.24016 19.3669 9.20526 19.3902 9.1471C19.4429 9.01371 19.4706 8.87172 19.4716 8.7283C19.4706 8.58487 19.4429 8.44288 19.3902 8.3095C19.3902 8.25133 19.332 8.21643 19.3088 8.15826C19.2782 8.09761 19.2432 8.03928 19.2041 7.98376L13.3874 1.00376C13.278 0.872443 13.1411 0.766836 12.9862 0.694452C12.8314 0.622068 12.6626 0.584686 12.4916 0.584963C12.2198 0.584432 11.9564 0.679099 11.7471 0.85253C11.6293 0.950191 11.5319 1.07013 11.4606 1.20548C11.3892 1.34083 11.3452 1.48893 11.3312 1.6413C11.3171 1.79367 11.3333 1.94732 11.3787 2.09344C11.4241 2.23955 11.4979 2.37528 11.5959 2.49283L15.8188 7.56496H2.02165C1.71311 7.56496 1.41721 7.68753 1.19905 7.9057C0.98088 8.12386 0.858315 8.41976 0.858315 8.7283C0.858315 9.03683 0.98088 9.33273 1.19905 9.5509C1.41721 9.76907 1.71311 9.89163 2.02165 9.89163Z"
          fill={color == 'primary' ? '#FFFFFF' : '#252160'}
        />
      </svg>
    )}
  </S.Wrapper>
);

export default Button;
