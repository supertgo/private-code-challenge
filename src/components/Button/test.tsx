import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helper';

import Button from '.';

describe('<Button />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<Button>children</Button>);

    expect(screen.getByText(/children/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small button', () => {
    renderWithTheme(<Button size="small">children</Button>);

    expect(screen.getByRole('button', { name: /children/i })).toHaveStyle({
      width: '13rem'
    });
  });

  it('should render the medium button', () => {
    renderWithTheme(<Button>children</Button>);

    expect(screen.getByRole('button', { name: /children/i })).toHaveStyle({
      width: '30rem'
    });
  });

  it('should render the small button', () => {
    renderWithTheme(<Button size="large">children</Button>);

    expect(screen.getByRole('button', { name: /children/i })).toHaveStyle({
      width: '38rem'
    });
  });

  it('should render the secundary color', () => {
    renderWithTheme(<Button color="secundary">children</Button>);

    expect(screen.getByRole('button', { name: /children/i })).toHaveStyle({
      backgroundColor: '#3AFFE5'
    });
  });

  it('should render the buttun with icon', () => {
    renderWithTheme(<Button icon>children</Button>);

    expect(screen.getByRole('button', { name: /children/i })).toHaveStyle({
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center'
    });
  });
});
