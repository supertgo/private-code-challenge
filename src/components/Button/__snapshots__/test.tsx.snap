// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the heading 1`] = `
.c0 {
  border: none;
  height: 4.3rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  text-align: center;
  font-size: 1.8rem;
  font-weight: 600;
  border-radius: 0.4rem;
  background: #252160;
  width: 30rem;
  color: #FFFFFF;
}

<button
  class="c0"
  color="primary"
>
  <span>
    children
  </span>
</button>
`;
